File: LeetCode75_LevelTwo/Binary_Tree_Right_Side_View/Binary_Tree_Right_Side_View.py
# 199. Binary Tree Right Side View Python
from collections import deque
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None: return []
        next_level = deque([root,])
        rightside = []

        while next_level:
            curr_level = next_level
            next_level = deque()

            while curr_level:
                node = curr_level.popleft()

                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
            
            rightside.append(node.val)
        return rightside

File: LeetCode75_LevelTwo/Binary_Tree_Right_Side_View/Binary_Tree_Right_Side_View.cc
// 199. Binary Tree Right Side View C++
#include <vector>
#include <queue>

//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    std::vector<int> output;
public:
    std::vector<int> rightSideView(TreeNode* root) {
        if (!root) return output;
        std::queue<TreeNode*> dq;
        dq.push(root);
        while (!dq.empty()) {
            int n = dq.size();
            for (int i = 0; i < n; i++) {
                TreeNode* node = dq.front();
                dq.pop();
                if (i == n - 1){
                    output.emplace_back(node->val);
                }
                if (node->left) dq.push(node->left);
                if (node->right) dq.push(node->right);
            }
        }
        return output;
    }
};
